### PR TITLE
For #1001: Enforce parameter order check

### PR DESCRIPTION
**Pull request by paulodamaso**
For #1001: Enforce parameter order check
- Implemented `JavadocParameterOrderCheck.java`
- Corrected tests for checking constructors too
- Added class to `checks.xml` and texts execution

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1012](https://github.com/teamed/qulice/issues/1012) is now in scope, role is `REV`

**Comment by 0crat** [comment]
@krzyk[/z](https://www.0crat.com/u/krzyk) everybody who has role `REV` is banned at [#1012](https://github.com/teamed/qulice/issues/1012); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by paulodamaso** [comment]
@krzyk ping

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java [hunk 1]
```diff
@@ -30,36 +30,216 @@
...
- *  signature. Implement this check and add its tests to checks.xml under
- *  custom checks.
  */
+@SuppressWarnings({"PMD.AvoidInstantiatingObjectsInLoops", "PMD.LongVariable"})
 public final class JavadocParameterOrderCheck extends AbstractCheck {
+
+    /**
+    * Compiled regexp to match Javadoc tags that take an argument.
+    */
+    private static final Pattern MATCH_JAVADOC_ARG = CommonUtil.createPattern(
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java]: @paulodamaso Why not `Pattern.compile`?

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java [hunk 2]
```diff
@@ -30,36 +30,216 @@
...
- * @todo #847:30min Enforce right parameter order in javadoc. Checkstyle must
- *  check if parameter order in javadoc is the same as is in the method
- *  signature. Implement this check and add its tests to checks.xml under
- *  custom checks.
  */
+@SuppressWarnings({"PMD.AvoidInstantiatingObjectsInLoops", "PMD.LongVariable"})
 public final class JavadocParameterOrderCheck extends AbstractCheck {
+
+    /**
+    * Compiled regexp to match Javadoc tags that take an argument.
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java]: @paulodamaso looks like your javadoc comments are not indented by one additional space, here it should look like:
```
/**
 * Text
 */
```

But it looks like:
```

/**
* Text
*/
```

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java [hunk 3]
```diff
@@ -30,36 +30,216 @@
...
+
+    /**
+     * Returns the param tags in a javadoc comment.
+     *
+     * @param comment The Javadoc comment
+     * @return The param tags found
+     */
+    private static List<JavadocTag> getMethodTags(final TextBlock comment) {
+        final String[] lines = comment.getText();
+        final List<JavadocTag> tags = new ArrayList<>(0);
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java]: @paulodamaso use LinkedList instead

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/JavadocParameterOrderCheck/Invalid.java [hunk 4]
```diff
@@ -9,6 +9,20 @@
  */
 public final class Invalid {
 
+    /**
+     * A field.
+     */
+    private final String field;
+
+    /**
+     * Javadoc with parameters in same order than the constructor signature.
```
Comment (krzyk) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/JavadocParameterOrderCheck/Invalid.java]: @paulodamaso `with parameters in different order ...`

**Comment by paulodamaso** [comment]
@krzyk Done, please take a look

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java [hunk 5]
```diff
@@ -30,36 +30,216 @@
...
- *  custom checks.
  */
+@SuppressWarnings({"PMD.AvoidInstantiatingObjectsInLoops", "PMD.LongVariable"})
 public final class JavadocParameterOrderCheck extends AbstractCheck {
+
+    /**
+     * Compiled regexp to match Javadoc tags that take an argument.
+     */
+    private static final Pattern MATCH_JAVADOC_ARG = Pattern.compile(
+        "^\\s*(?>\\*|\\/\\*\\*)?\\s*@(param)\\s+(\\S+)\\s+\\S*", 0
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java]: @paulodamaso the `flags = 0` is default for single argument Pattern.compile

Comment (paulodamaso) [reply qulice-checkstyle/src/main/java/com/qulice/checkstyle/JavadocParameterOrderCheck.java]: @krzyk Commited using no parameters, let's see what happens

**Comment by 0crat** [comment]
Job `gh:teamed/qulice#1012` is not assigned, can't get performer

**Comment by 0crat** [comment]
The job [#1012](https://github.com/teamed/qulice/issues/1012) is now out of scope

**Pull request merged**